gobblin-service/src/main/java/org/apache/gobblin/service/modules/flow/LoadBasedFlowEdgeImpl.java

```diff
@@ -0,0 +1,160 @@
+  @Getter
+  private ServiceNode targetNode;
+  @Getter
+  private SpecExecutor specExecutorInstance;
+
+  private final FlowEdgeProps _flowEdgeProps;
+
+  /**
+   * {@link #edgeSecurity}'s value comes from {@link FlowEdgeProps} initially
+   * and can be overrided by {@link #setEdgeSecurity(boolean)} and
+   * {@link org.jgrapht.graph.DirectedWeightedMultigraph#setEdgeWeight(Object, double)} method afterwards.
+   */
+  private boolean edgeSecurity;
+
+
+  public LoadBasedFlowEdgeImpl(ServiceNode sourceNode, ServiceNode targetNode,
+      FlowEdgeProps flowEdgeProps, SpecExecutor specExecutorInstance) {
+    this.sourceNode = sourceNode ;
+    this.targetNode = targetNode ;
+    this._flowEdgeProps = flowEdgeProps;
+    this.specExecutorInstance = specExecutorInstance;
+    this.edgeSecurity = flowEdgeProps.getInitialEdgeSafety();
+  }
+
+  public LoadBasedFlowEdgeImpl(ServiceNode sourceNode, ServiceNode targetNode,
+      SpecExecutor specExecutor){
+    this(sourceNode, targetNode, new FlowEdgeProps(ConfigUtils.configToProperties(specExecutor.getAttrs())),
+        specExecutor);
+  }
+
+  public double getEdgeLoad(){
+    return getWeight();
+  }
+
+  @Override
+  public String getEdgeIdentity(){
+    return this.calculateEdgeIdentity(this.sourceNode, this.targetNode, this.specExecutorInstance);
+  }
+
+  /**
+   * Since {@link #_flowEdgeProps} is supposed to be read-only once instantiated, set method here
+   * will overwrite the value in {@link #edgeSecurity}.
+   */
+  public void setEdgeSecurity(boolean security) {
+    this.edgeSecurity = security;
+  }
+
+  @Override
+  public Properties getEdgeProperties() {
+    return this._flowEdgeProps.getProperties();
+  }
+
+  @Override
+  /**
+   * Naive implementation for here.
+   */
+  public boolean isEdgeValid() {
+    return this.edgeSecurity;
+  }
+
+  /**
+   * A naive implementation of edge identity calculation.
+   * @return
+   */
+  public static String calculateEdgeIdentity(ServiceNode sourceNode, ServiceNode targetNode, SpecExecutor specExecutorInstance){
+    return sourceNode.getNodeName() + "-" + specExecutorInstance.getUri() + "-" + targetNode.getNodeName();
+  }
+
+  /**
+   * Recall that we need a triplet to uniquely define a {@link FlowEdge}:
+   * - {@link ServiceNode} sourceNode
+   * - {@link ServiceNode} targetNode
+   * - {@link SpecExecutor} SpecExecutor
+   *
+   * We DO NOT distinguish between two edges by other props like weight,
+   * as the load should be an attribute of an edge.
+   * These are IntelliJ-generated methods for equals and hashCode().
+   *
+   * @param o The object that being compared
+   * @return If two {@link LoadBasedFlowEdgeImpl} are equivalent.
+   */
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    LoadBasedFlowEdgeImpl that = (LoadBasedFlowEdgeImpl) o;
+
+    if (!sourceNode.equals(that.sourceNode)) {
```

Comment:
Do you need to implement equals for the fields here or is object reference compare the intended check?

Reply:
This is the intended. The comparison of an edge should include comparison of endpoint plus comparison of specExecutor. 

What I missed here is the `equals` method overriding in `ServiceNode`'s implementation which should not include comparison of node properties. Also in the `SpecExecutor`'s implementation, two `SpecExecutor`s with the same URI should be considered as the same `SpecExecutor`.